File: backend/models/project.py
"""
プロジェクトモデル

足場プロジェクトのデータモデルを定義
"""

from pydantic import Field, field_validator
from typing import Optional, Dict, Any, List
from datetime import date
from enum import Enum

from models.base import BaseDBModel


class ProjectStatus(str, Enum):
    """
    プロジェクトのステータス定義

    - draft: 下書き（作成中）
    - active: 進行中
    - completed: 完了
    - archived: アーカイブ済み
    """
    DRAFT = "draft"
    ACTIVE = "active"
    COMPLETED = "completed"
    ARCHIVED = "archived"


class ProjectBase(BaseDBModel):
    """
    プロジェクトの基本情報

    全てのプロジェクト関連モデルで共通するフィールド
    """
    name: str = Field(
        ...,
        description="プロジェクト名",
        min_length=1,
        max_length=200
    )
    description: Optional[str] = Field(
        default=None,
        description="プロジェクトの説明",
        max_length=1000
    )
    status: ProjectStatus = Field(
        default=ProjectStatus.DRAFT,
        description="プロジェクトのステータス"
    )
    customer_name: Optional[str] = Field(
        default=None,
        description="顧客名",
        max_length=100
    )
    site_address: Optional[str] = Field(
        default=None,
        description="現場住所",
        max_length=300
    )
    start_date: Optional[date] = Field(
        default=None,
        description="プロジェクト開始日"
    )
    end_date: Optional[date] = Field(
        default=None,
        description="プロジェクト終了日"
    )
    metadata: Optional[Dict[str, Any]] = Field(
        default=None,
        description="追加メタデータ（JSON形式）"
    )

    @field_validator('end_date')
    @classmethod
    def validate_end_date(cls, v: Optional[date], info) -> Optional[date]:
        """
        終了日が開始日より後であることを検証

        Args:
            v: 終了日
            info: バリデーション情報（start_dateを含む）

        Returns:
            Optional[date]: 検証済みの終了日

        Raises:
            ValueError: 終了日が開始日より前の場合
        """
        start_date = info.data.get('start_date')
        if v is not None and start_date is not None:
            if v < start_date:
                raise ValueError('終了日は開始日より後である必要があります')
        return v


class ProjectCreate(ProjectBase):
    """
    プロジェクト作成時のリクエストモデル

    新規プロジェクト作成に必要なフィールド
    """
    user_id: str = Field(
        ...,
        description="プロジェクト所有者のユーザーID（UUID）"
    )


class ProjectUpdate(BaseDBModel):
    """
    プロジェクト更新時のリクエストモデル

    全てのフィールドがオプショナル（部分更新可能）
    """
    name: Optional[str] = Field(
        default=None,
        description="プロジェクト名",
        min_length=1,
        max_length=200
    )
    description: Optional[str] = Field(
        default=None,
        description="プロジェクトの説明",
        max_length=1000
    )
    status: Optional[ProjectStatus] = Field(
        default=None,
        description="プロジェクトのステータス"
    )
    customer_name: Optional[str] = Field(
        default=None,
        description="顧客名",
        max_length=100
    )
    site_address: Optional[str] = Field(
        default=None,
        description="現場住所",
        max_length=300
    )
    start_date: Optional[date] = Field(
        default=None,
        description="プロジェクト開始日"
    )
    end_date: Optional[date] = Field(
        default=None,
        description="プロジェクト終了日"
    )
    metadata: Optional[Dict[str, Any]] = Field(
        default=None,
        description="追加メタデータ（JSON形式）"
    )


class Project(ProjectBase):
    """
    プロジェクトの完全なデータモデル

    データベースから取得したプロジェクト情報を表現
    """
    user_id: str = Field(
        ...,
        description="プロジェクト所有者のユーザーID（UUID）"
    )


class ProjectListResponse(BaseDBModel):
    """
    プロジェクト一覧取得時のレスポンスモデル
    """
    projects: List[Project] = Field(
        ...,
        description="プロジェクトのリスト"
    )
    total: int = Field(
        ...,
        description="総プロジェクト数"
    )
    page: int = Field(
        ...,
        description="現在のページ番号"
    )
    page_size: int = Field(
        ...,
        description="1ページあたりのアイテム数"
    )


class ProjectDuplicateRequest(BaseDBModel):
    """
    プロジェクト複製時のリクエストモデル
    """
    new_name: Optional[str] = Field(
        default=None,
        description="複製後のプロジェクト名（指定しない場合は「〜のコピー」が付与される）",
        max_length=200
    )

File: frontend/src/types/project.ts
/**
 * プロジェクト型定義
 *
 * バックエンドのプロジェクトモデルと対応する型定義
 */

/**
 * プロジェクトステータス
 */
export type ProjectStatus = 'draft' | 'active' | 'completed' | 'archived'

/**
 * プロジェクトの基本情報
 */
export interface Project {
  id: string
  user_id: string
  name: string
  description?: string | null
  status: ProjectStatus
  customer_name?: string | null
  site_address?: string | null
  start_date?: string | null // ISO 8601形式の日付文字列
  end_date?: string | null   // ISO 8601形式の日付文字列
  metadata?: Record<string, unknown> | null
  created_at: string // ISO 8601形式のタイムスタンプ
  updated_at: string // ISO 8601形式のタイムスタンプ
}

/**
 * プロジェクト作成リクエスト
 */
export interface ProjectCreateRequest {
  name: string
  description?: string | null
  status?: ProjectStatus
  customer_name?: string | null
  site_address?: string | null
  start_date?: string | null
  end_date?: string | null
  metadata?: Record<string, unknown> | null
  user_id?: string // バックエンドで上書きされるため、フロントエンドからは不要
}

/**
 * プロジェクト更新リクエスト
 */
export interface ProjectUpdateRequest {
  name?: string
  description?: string | null
  status?: ProjectStatus
  customer_name?: string | null
  site_address?: string | null
  start_date?: string | null
  end_date?: string | null
  metadata?: Record<string, unknown> | null
}

/**
 * プロジェクト一覧レスポンス
 */
export interface ProjectListResponse {
  projects: Project[]
  total: number
  page: number
  page_size: number
}

/**
 * プロジェクト複製リクエスト
 */
export interface ProjectDuplicateRequest {
  new_name?: string | null
}

/**
 * ステータスラベルマッピング
 */
export const PROJECT_STATUS_LABELS: Record<ProjectStatus, string> = {
  draft: '未着手',
  active: '進行中',
  completed: '完了',
  archived: 'アーカイブ',
}

/**
 * ステータスアイコンマッピング
 */
export const PROJECT_STATUS_ICONS: Record<ProjectStatus, string> = {
  draft: '📝',
  active: '🚀',
  completed: '✅',
  archived: '📦',
}

/**
 * ステータスカラーマッピング（Tailwind CSS）
 */
export const PROJECT_STATUS_COLORS: Record<
  ProjectStatus,
  { bg: string; text: string; border: string }
> = {
  draft: {
    bg: 'bg-indigo-100/80 dark:bg-indigo-900/30',
    text: 'text-indigo-900 dark:text-indigo-300',
    border: 'border-indigo-600 dark:border-indigo-600',
  },
  active: {
    bg: 'bg-cyan-100/80 dark:bg-cyan-900/30',
    text: 'text-cyan-900 dark:text-cyan-300',
    border: 'border-cyan-600 dark:border-cyan-600',
  },
  completed: {
    bg: 'bg-emerald-100/70 dark:bg-emerald-900/30',
    text: 'text-emerald-900 dark:text-emerald-300',
    border: 'border-emerald-600 dark:border-emerald-600',
  },
  archived: {
    bg: 'bg-gray-100/80 dark:bg-gray-900/30',
    text: 'text-gray-900 dark:text-gray-300',
    border: 'border-gray-600 dark:border-gray-600',
  },
}
